### PR TITLE
Fix fromjson() to support reading from stdin

### PR DESCRIPTION
**Pull request by yaniv-aknin**
Trivial PR to support reading from stdin (`source=None`) in `fromjson()`.

Two things you may want me to do before merging -

1. I didn't add tests because there are no tests I could find for the use of `source=None` in any `io` module (e.g., also not in `csv` etc). Generally, perhaps I missed them but I didn't see tests that fork/exec the `bin/petl` executable at all...?
2. Add `source=None` to other modules. I ran `git grep 'def from' | grep source | grep -v source=None` and found a few more that miss it, let me know if you'd like me to add it to any/all of them.
    Modules which appear to be missing `source=None`: `avro`, `bcolz`, `pytables` (maybe), `xml`

## Checklist

Use this checklist to ensure the quality of pull requests that include new code and/or make changes to existing code.

* [ ] Source Code guidelines:
  * [ ] Includes unit tests
  * [ ] New functions have docstrings with examples that can be run with doctest
  * [ ] New functions are included in API docs
  * [ ] Docstrings include notes for any changes to API or behavior
  * [ ] All changes are documented in docs/changes.rst
* [ ] Versioning and history tracking guidelines:
  * [ ] Using atomic commits whenever possible
  * [ ] Commits are reversible whenever possible
  * [ ] There are no incomplete changes in the pull request
  * [ ] There is no accidental garbage added to the source code
* [ ] Testing guidelines:
  * [ ] Tested locally using `tox` / `pytest`
  * [ ] Rebased to `master` branch and tested before sending the PR
  * [ ] Automated testing passes (see [CI](https://github.com/petl-developers/petl/actions))
  * [ ] Unit test coverage has not decreased (see [Coveralls](https://coveralls.io/github/petl-developers/petl))
* [ ] State of these changes is:
  * [ ] Just a proof of concept
  * [ ] Work in progress / Further changes needed
  * [ ] Ready to review
  * [ ] Ready to merge

### CODE REVIEW

File: petl/io/json.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 from petl.util.base import data, Table, dicts as _dicts, iterpeek
 
 
-def fromjson(source, *args, **kwargs):
+def fromjson(source=None, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (juarezr) [reply petl/io/json.py:19]: >  Code scanning / Prospector (reported by Codacy) 
>
> Keyword argument before variable positional arguments list in the definition of fromjson function (keyword-arg-before-vararg)

This wouldn't break existing code?

Maybe [read_source_from_arg](https://github.com/petl-developers/petl/blob/master/petl/io/sources.py#L465)  can be taught to read even when stdin is passed as argument as in:

``` python
    table1 = etl.fromjson(sys.stdin, header=["foo", "bar"])
    print(table1)
```

[the remark this answers is not in the document]
Comment (yaniv-aknin) [reply petl/io/json.py:19]: I might be wrong, but I don't think it should break existing code.

All these assertions pass -

```python
>>> def f(s, *a, **kw):
...     return (s, a, kw)
... 
>>> def fn(s=None, *a, **kw):
...     return (s, a, kw)
... 
>>> assert f("hello") == fn("hello")
>>> assert f(None) == fn(None)
>>> assert f("hello", 1, 2) == fn("hello", 1, 2)
>>> assert f("hello", k=1) == fn("hello", k=1)
>>> assert f(s="hello") == fn(s="hello")
>>> assert f("hello", 1, 2, k=1, j=2) == fn("hello", 1, 2, k=1, j=2)
```

re. `read_source_from_arg()` - my interest is in the `petl` executable, i.e., I would love for syntax like this to work in shell: `echo '{"a":2, "b":4}' | petl 'fromjson(lines=True)'`.

What do you think?

[the remark this answers is not in the document]
Comment (juarezr) [reply petl/io/json.py:19]: I think that:

1. Using petl as a command line tool is an interesting use case and we should give it some love by:
   a. Adding more examples in the [documentation](https://petl.readthedocs.io/en/stable/intro.html#petl-executable).
   b. Mention this usage pattern in the repository Readme/Frontage
2. Currently [fromjson()](https://petl.readthedocs.io/en/stable/io.html#petl.io.json.fromjson) source argument is not consistent  with other functions like:
   a.  [tojson()](https://petl.readthedocs.io/en/stable/io.html#petl.io.json.tojson) and [tojsonarrays()](https://petl.readthedocs.io/en/stable/io.html#petl.io.json.tojsonarrays)
   b. [fromcsv()](https://petl.readthedocs.io/en/stable/io.html#petl.io.csv.fromcsv)
3. We value not breaking existing code.
   a. But I haven't had the time to check if this is the case yet.
   b. If not it will be a good change for API consistency
   c. I would love to hear more opinions about this change.

**Comment by juarezr** [comment]
Adding to the investigation:

The [fromjson()](https://github.com/petl-developers/petl/blob/master/petl/io/json.py#L19) parameters **args** and **kwargs** are forwarded to the [json library call](https://github.com/petl-developers/petl/blob/master/petl/io/json.py#L116):

``` python
def fromjson(source, *args, **kwargs):
    """..."""
   ...

...
dicts = json.load(f, *self.args, **self.kwargs)
...
```

**Comment by juarezr** [comment]
But the [json library calls](https://docs.python.org/3/library/json.html#json.load) used take only one positional argument followed by others keyword arguments:

``` python
 def json.load(fp, *, cls=None, object_hook=None, parse_float=None, parse_int=None, parse_constant=None, object_pairs_hook=None, **kw):
    ...

def  json.loads(s, *, cls=None, object_hook=None, parse_float=None, parse_int=None, parse_constant=None, object_pairs_hook=None, **kw):
   ...
```

It looks like the **args** parameter is simply ignored by **json.load** and  **json.loads** calls.

Now I'm wondering if the function syntax shouldn't be:

``` json
def fromjson(source=None, *, **kwargs):
    """..."""
   ...

**Comment by juarezr** [comment]
@yaniv-aknin,

After looking at the source code, I've concluded that [your proposal](https://github.com/petl-developers/petl/pull/667#issue-2262715960) looks good.

Also, some considerations:

- It seems like some functions are not ready to use stdin or even can't work with stdin at all. Notice that :
  -  [pytables](https://petl.readthedocs.io/en/stable/io.html#petl.io.pytables.fromhdf5) function [fromhdf5](https://github.com/petl-developers/petl/blob/master/petl/io/pytables.py#L135)
  - [bcolz](https://petl.readthedocs.io/en/stable/io.html#petl.io.bcolz.frombcolz) function [frombcolz](https://github.com/petl-developers/petl/blob/master/petl/io/bcolz.py#L84)
- As [fromjson](https://github.com/petl-developers/petl/blob/master/petl/io/json.py#L19) receives, but doesn't use the **args** parameter, maybe it would be interesting to change the function syntax/signature as described [above](https://github.com/petl-developers/petl/pull/667#issuecomment-2081655742).
- This would be a good case for adding some test cases to avoid future regressions as the test suite is the major guarantee of stability for [petl](https://coveralls.io/github/petl-developers/petl?branch=master). The test cases will ensure that petl executable won't stop working after new features are merged.

So, what are your plans for this PR?

**Comment by yaniv-aknin** [comment]
Thanks @juarezr , this sounds good.

1. I'd be happy to add `source=None` to more cases, where it'd work easily. Where it'd take a more substantial refactoring (or even just infeasible), I'd leave it unchanged.

2. I'm mildly indifferent about changing the signature, but don't mind doing it of course. If `(sources=None, *, **kwargs)` looks good to you, great, I'll do that.

3. I'm also happy to add some tests that fork/exec `petl`, and in particular will test the "pipe from stdin" case.
   I propose to - 
    * Create a  new test module `petl/test/test_script.py`
    * Have it find `bin/petl` in the package directory and fork/exec it using the `subprocess` module, testing results
    * Add a few simple test cases around that
    * Make the above pass in `petl`'s supported environments (2.7, 3.6-3.12)

Does that sound reasonable and I can update the PR?

**Comment by juarezr** [comment]
After checking it, I've found that the change in signature won't work:

``` python
>>> def fromjson(source=None, *, **kwargs):
...    print("source:", source, "kwargs:", kwargs) 
... 
  File "<stdin>", line 1
SyntaxError: named arguments must follow bare *
```

So we should keep the **args** argument as it is currently, and proceed.

Eventually, we could remove the **self.args** property and the forwarding to **json.load** functions as it is pointless in:

``` python
dicts = json.load(f, *self.args, **self.kwargs)
```

But it shouldn't be mandatory.

**Comment by yaniv-aknin** [comment]
I've made some progress, but not all checks pass yet.

* Python2.7 -- probably my fault, I need to look further
* Windows tests -- not sure, if I can sort out a windows test environment I will do it
* macOS tests -- seem broken unrelated to me
* security test complaining about shell=True -- I disagree with the check, I am literally testing shell behaviour; I could do this differently, but this is short readable and (in test context) perfectly safe
* No docstring - don't think they're needed for something like this...?

I'll try to look into this again next week, but admit I'm beginning to question how much time I can put into this; small fix, lots of overhead :sweat_smile:

**Comment by juarezr** [comment]
@CodiumAI-Agent /review
